### PR TITLE
Experiment: fmt::Arguments as closure

### PR DESCRIPTION
**Pull request by m-ou-se**
This is part of https://github.com/rust-lang/rust/issues/99012

This implements the "closure idea" for fmt::Arguments.

For now it uses a simple `enum { Fn(&'a dyn Fn), StaticStr(&'static str) }` to handle the static str case for `as_str()`. A slightly more optimized version could reduce the size of `fmt::Arguments` to two pointers with some tricks.

Includes https://github.com/rust-lang/rust/pull/100996 and https://github.com/rust-lang/rust/pull/101569

With this, `println!("Hello, {0} {0:x} {0:#x}!", 100)` expands to:

```rust
_print(Arguments::new(&match (&100,) {
    _args => |w: &mut dyn Write| -> Result {
        w.write_str("Hello, ")?;
        Display::fmt(_args.0, &mut Formatter::new(w))?;
        w.write_str(" ")?;
        LowerHex::fmt(_args.0, &mut Formatter::new(w))?;
        w.write_str(" ")?;
        LowerHex::fmt(
            _args.0,
            &mut Formatter::new(w)
                .with_options(4u32, ' ', Alignment::Unknown, None, None)
        )?;
        w.write_str("!\n")?;
        Ok(())
    },
}))
```

### CODE REVIEW

**Comment by joshtriplett** [comment]
At the risk of microoptimization: might it be worth recognizing the *extremely* common case of a single-character `write_str` and turning it into a call to `Write::write_char(f, ' ')` (which could potentially be marked `#[inline]` so that it directly invokes `write_char` on `f.buf`)? Would that potentially be a net win?

**Comment by joshtriplett** [comment]
I'm incredibly excited about this, and in particular I'm excited for being able to give the optimizer more optimization opportunities.

**Comment by m-ou-se** [comment]
@joshtriplett `write_char`'s default implementation is `self.write_str(c.encode_utf8(&mut [0; 4]))`, so I'm not sure if that actually helps.

**Comment by m-ou-se** [comment]
> The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8247427157?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/05fefb7498ddd8d5a0cedaca8d0e26386360297f/checks/8247427157/logs)
> Click to see the possible cause of the failure (guessed by this bot)

```diff
-                ╾───────ALLOC_ID───────╼ 01 00 00 00 00 00 00 00 │ ╾──────╼........
+                ╾──────ALLOC_ID───────╼ 01 00 00 00 00 00 00 00 │ ╾──────╼........
```

Uhh.. How did that happen? Somehow an off-by-one error when formatting `"╾{0:─^1$}╼"`? :S I can't seem to reproduce this locally.

**Comment by joshtriplett** [comment]
On Tue, Sep 13, 2022 at 03:16:47AM -0700, Mara Bos wrote:
> @joshtriplett `write_char`'s default implementation is `self.write_str(c.encode_utf8(&mut [0; 4]))`, so I'm not sure if that actually helps.

Ah. I was hoping that with sufficient inlining it would turn into "just
append this byte".

Nevermind, then.

- Josh Triplett

**Comment by m-ou-se** [comment]
I've added some #[inline] attributes to the formatting functions in `core`. Now, calls to `write!()` can be fully understood by the optimzer:

For example:

```rust
use std::fmt::{self, Write};

struct Thing;

extern "C" {
    fn write(fd: i32, buf: *const u8, count: usize) -> i32;
}

impl fmt::Write for Thing {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        unsafe { write(1, s.as_ptr(), s.len()) };
        Ok(())
    }
}

pub fn test() {
    let _ = write!(Thing, "Hello, {}!", "world");
}
```

This compiles down to basically `write("Hello, "); write("world"); write("!");`. :tada: :confetti_ball:

Previously, this would include the full `Display for str` implementation including the `Formatter::pad` function with all its functionality, and the full `fmt::write` function, and construct a `fmt::Arguments` on the stack, etc.

```asm
	push rbx
	lea rsi, [rip + .L__unnamed_1] ; "Hello, "
	mov rbx, qword ptr [rip + write@GOTPCREL]
	mov edx, 7
	mov edi, 1
	call rbx

	lea rsi, [rip + .L__unnamed_2]; "world"
	mov edx, 5
	mov edi, 1
	call rbx

	lea rsi, [rip + .L__unnamed_3]; "!"
	mov edx, 1
	mov edi, 1
	mov rax, rbx
	pop rbx
	jmp rax
```

**Comment by m-ou-se** [comment]
The closure now takes a `&mut Formatter`, which means that when optimizing the closure without any context (e.g. behind a non-inlined function), the compiler can't assume anything about the state of that `Formatter`. (Are any non-default formatting flags set?) If it isn't inlining a Display implementation, it might also assume that the Display::fmt call might have changed the formatting flags, preventing optimization of the next Display::fmt call.

To fix this, I'm now changing the closure signature to take a `&mut dyn Write` instead, and constructing a new `Formatter` for every `Display::fmt` call.

**Comment by bjorn3** [comment]
Flags like `#` can leak in from the outer `format_args!()` into the inner `format_args!()`, right? That is `format!("{:#?}", val)` would set the `#` flag for `write!()` call inside the `Debug` impl for `val`. Wouldn't reconstructing the `Formatter` every time break this? Or did I misremember the cases where flags get passed in to child formatting?

**Comment by m-ou-se** [comment]
@bjorn3 `<Debug for ..>::fmt` gets a `Formatter` and might pass that formatter on to other `<Debug for ..>::fmt` calls (or other traits than `Debug`, like `Display`), so it gets the same flags. A nested `write!()` or `format_args!()` call would only get that `Formatter` as a `&impl Write`, around which it will construct a new `Formatter`, making it unaffected by any flags on the outermost one.

In other words: Formatters get 'inherited' (passed) between nested Display/Debug/...::fmt calls, but not between "nested" format_args!()/write!() calls.

**Comment by m-ou-se** [comment]
Okay let's try a perf run..

I'm not expecting great results, because this just tests rustc itself. The approach in this PR results in a closure per format_args!(), meaning more codegen and more work for the optimizer.

We'll need separate benchmarking for runtime performance.

The try build will also allow folks to easily test out this implementation using [rustup-toolchain-install-master](https://github.com/kennytm/rustup-toolchain-install-master).

@bors try @rust-timer queue

**Comment by m-ou-se** [comment]
Yup, that's exactly as expected. ^^

The `incr-patched: println` ones go up a lot, since compiling a format string is now more expensive. Similarly `full` for crates with a lot of formatting also go up significantly, such as `cargo` by about 23%.

There might be ways to improve the compiler performance, but for now I'm not going to address this. The purpose of this experimental PR for now is to see what the effect of such an approach can be on runtime performance and binary size.

Only if we conclude this might be the best approach for runtime performance and binary size, is it worth looking into compiler performance.

**Comment by m-ou-se** [comment]
Interestingly, some of the `doc` jobs show a significant improvement, specifically in `render_item`. That might be a sign that it can indeed have a very positive impact on runtime performance. :)

**Comment by Dylan-DPC** [comment]
Closing this as it was an experiment.
